homeassistant/components/habitica/calendar.py

```diff
@@ -0,0 +1,179 @@
+        )
+
+        return events_sorted[0] if events_sorted else None
+
+    async def async_get_events(
+        self, hass: HomeAssistant, start_date: datetime, end_date: datetime
+    ) -> list[CalendarEvent]:
+        """Return calendar events within a datetime range."""
+
+        return [
+            CalendarEvent(
+                start=start,
+                end=start + timedelta(days=1),
+                summary=task["text"],
+                description=task["notes"],
+                uid=task["id"],
+            )
+            for task in self.coordinator.data.tasks
+            if task["type"] == HabiticaTaskType.TODO
+            and not task["completed"]
+            and task.get("date")
+            and (start := to_date(task["date"]))
+            and (start_date.date() <= start <= end_date.date())
+        ]
+
+
+class HabiticaDailiesCalendarEntity(HabiticaCalendarEntity):
+    """Habitica dailies calendar entity."""
+
+    entity_description = CalendarEntityDescription(
+        key=HabiticaCalendar.DAILIES,
+        translation_key=HabiticaCalendar.DAILIES,
+    )
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the next upcoming event."""
+
+        today = datetime.combine(
+            dt_util.now().date(), time(), dt_util.DEFAULT_TIME_ZONE
+        )
+        events = [
+            CalendarEvent(
+                start=(start := next_recurrence.date()),
+                end=start + timedelta(days=1),
+                summary=task["text"],
+                description=task["notes"],
+                uid=task["id"],
+            )
+            for task in self.coordinator.data.tasks
+            if task["type"] == HabiticaTaskType.DAILY
+            and not task["completed"]
+            and task["everyX"]
+            if (next_recurrence := build_rrule(task).after(today, inc=True))
+        ]
+        events_sorted = sorted(
+            events,
+            key=lambda event: (
+                event.start,
+                self.coordinator.data.user["tasksOrder"]["dailys"].index(event.uid),
+            ),
+        )
+
+        return events_sorted[0] if events_sorted else None
+
+    async def async_get_events(
+        self, hass: HomeAssistant, start_date: datetime, end_date: datetime
+    ) -> list[CalendarEvent]:
+        """Return calendar events within a datetime range."""
+
+        today = to_date(self.coordinator.data.user["lastCron"])
+        # returns only todays and future dailies.
+        # If a daily is completed it will not be shown for today but still future recurrences
+        # If the cron hasn't run, not completed dailies are yesterdailies and displayed yesterday
+        return [
+            CalendarEvent(
+                start=start,
+                end=start + timedelta(days=1),
+                summary=task["text"],
+                description=task["notes"],
+                uid=task["id"],
+                rrule=str(build_rrule(task))[30:],
```

Comment:
This is stripping  the dtstart part of the rrule. Otherwise the calendar is unable to parse the rule. I was looking for a method to get a rule from rrule without the dtstart part but couldn't find one

Reply:
Two suggestions:
- I would suggest extracting a descriptive function name that describes what this is doing. (I happy to know the format this library produces includes dtstart because i've read that library code before, but not sure that is obvious to other readers)
- I think there may be a newline in there as a delimiter. Explicitly checking the rrule prefix would help too.  Adding tests would also be useful.